include/seqan3/alphabet/concept.hpp

```diff
@@ -16,22 +16,936 @@
+#include <seqan3/alphabet/exception.hpp>
 #include <seqan3/std/concepts>
 
+// ============================================================================
+// to_rank()
+// ============================================================================
+
+namespace seqan3::detail::adl::only
+{
+
+//!\brief Call `to_rank()` on the argument, but only consider ADL.
+//!\ingroup alphabet
+template <typename alph_t>
+//!\cond
+    requires requires { to_rank(std::declval<alph_t>()); }
+//!\endcond
+constexpr decltype(auto) to_rank_(alph_t && a) noexcept(noexcept(to_rank(std::declval<alph_t>())))
+{
+    return to_rank(std::forward<alph_t>(a));
+}
+
+} // namespace seqan3::detail::adl::only
+
+namespace seqan3::detail
+{
+
+// Auxiliary concepts
+//!\cond
+template <typename alph_t>
+SEQAN3_CONCEPT has_to_rank_adl_overload = requires { { adl::only::to_rank_(std::declval<alph_t>()) }; };
+
+template <typename alph_t>
+SEQAN3_CONCEPT has_to_rank_adaptation_overload = requires { { ::seqan3::adaptation::to_rank(std::declval<alph_t>()) }; };
+
+template <typename alph_t>
+SEQAN3_CONCEPT has_to_rank_member = requires { { std::declval<alph_t>().to_rank() }; };
+//!\endcond
+
+//!\brief Functor definition for seqan3::to_rank.
+struct to_rank_fn
+{
+    //!\brief Operator definition.
+    template <typename alph_t>
+    //!\cond
+        requires has_to_rank_adl_overload<alph_t>        ||
+                 has_to_rank_adaptation_overload<alph_t> ||
+                 has_to_rank_member<alph_t>
+    //!\endcond
+    constexpr decltype(auto) operator()(alph_t const a) const noexcept
+    {
+        if constexpr (has_to_rank_adl_overload<alph_t>)
+        {
+            static_assert(noexcept(adl::only::to_rank_(a)),
+                          "Only overloads that are marked noexcept are picked up by seqan3::to_rank().");
+            return adl::only::to_rank_(a);
+        }
+        else if constexpr (has_to_rank_adaptation_overload<alph_t>)
+        {
+            static_assert(noexcept(::seqan3::adaptation::to_rank(a)),
+                          "Only overloads that are marked noexcept are picked up by seqan3::to_rank().");
+            return ::seqan3::adaptation::to_rank(a);
+        }
+        else
+        {
+            static_assert(noexcept(a.to_rank()),
+                          "Only members that are marked noexcept are picked up by seqan3::to_rank().");
+            return a.to_rank();
+        }
+    }
+};
+
+} // namespace seqan3::detail
+
+namespace seqan3
+{
+
+/*!\name Function objects
+ * \{
+ */
+
+/*!\brief Return the rank representation of an (Semi-)Alphabet object.
+ * \param alph The target object; its type must model seqan3::Alphabet.
+ * \returns The rank representation; an integral type.
+ * \ingroup alphabet
+ * \details
+ *
+ * This is a function object. Invoke it with the parameters specified above.
+ *
+ * ### Example
+ *
+ * ```cpp
+ * char d = seqan3::to_char('A'_dna5);                  // d == 'A'
```

Comment:
should be a to_rank example?

Reply:
But, I have changed that already, or not?